port mouse settings from mpc-be

adipose opened the pull request:
This is a work in progress.  I didn't update the themed elements yet, but the basics work well enough for testing.  We will have to decide about the following deficiencies:

1. No differentiation between mouse down and mouse up
2. Inability to use one mouse action for multiple commands (maybe a good thing)
3. Not porting the existing commands from the old table.  Could be fixed.

adipose committed:
port mouse settings from mpc-be

adipose committed:
.

adipose commented:
4. Supports "right click" as a modifier to other clicks
5. No support for ALT modifier

clsid2 commented:
Can you rebase and make a test build?

adipose commented:
https://mega.nz/file/NRYTiADR#D9nVki_nNqGuGt8MpcwbpkG_AtPb4lR7NcRw-C-sRfo

Sorry, forgot to upload this build.

xLn2 commented:
Just some ideas:
1. Instead of Single "Default" button, there may be "Select All" + "Reset Selected" buttons, or single "Reset Selected" button
2. Calculating drowdown menu width // It is calculation seems like fine but doesn't consider scrollbar overlaps it (maybe order of things related). Absence of scrollbar might solve the issue, but I don't know future use cases here. This is also happening on Playback>Fullscreen and Pininfo window. 
![resim](https://github.com/clsid2/mpc-hc/assets/3073922/96c97b43-144d-49f3-83f8-b198814795be)
3. Theming and default width size of "Right button".

Thanks.

Just a quick report about fuctionality:
Ctrl, Shift and Right button modifiers looks like completely unfunctioning.
Disabling "Move window by the video area" setting doesn't have affect.
Disabling "If nothing is loaded, open recent file" setting doesn't have affect.

adipose commented:
I have not made any effort to theme the elements of the mouse settings page.  Right now, only functionality should be tested.

clsid2 commented:
I will get back to this after 2.3.2 release (soon).

clsid2 commented:
I think the "if nothing loaded" option can be removed. That is simply what play action does (when assigned to left click).

The "move window by video area" can also be removed. If there are really people who need to disable that, then such an obscure option is better placed in the advanced settings. So general settings pages stay clean.

Then the "left button" surrounding box could be removed and the remaining two labels changed to "left click" and "left double click" (plus making them wider for translations).

The table for the other keys might be a bit confusing for some at first for how extra the modifiers work. I suggest changing the table headers to "+ Ctrl" and "+ Shift" and "+ Right button".

There is plenty of combos possible, so no need for ALT modifier.

adipose committed:
merge upstream

adipose committed:
remove unimportant mouse options; fix themed elements

adipose commented:
> I think the "if nothing loaded" option can be removed. That is simply what play action does (when assigned to left click).
 
Removed.

> The "move window by video area" can also be removed. If there are really people who need to disable that, then such an obscure option is better placed in the advanced settings. So general settings pages stay clean.

Removed.

adipose committed:
align widgets better; add plus to column headings

adipose commented:
> Then the "left button" surrounding box could be removed and the remaining two labels changed to "left click" and "left double click" (plus making them wider for translations).

I turned it into a groupbox for the standard mouse buttons.
 
> The table for the other keys might be a bit confusing for some at first for how extra the modifiers work. I suggest changing the table headers to "+ Ctrl" and "+ Shift" and "+ Right button".

Add "+" to them.  Note, the plus is not translatable--not sure if we care.  Ctrl/Shift were already hardcoded but "right click" was translatable.  If important, I can add proper translatable strings.

adipose commented:
![image](https://github.com/user-attachments/assets/bcea4938-b132-4149-9a6e-177bd6adda5a)

xLn2 commented:
![resim](https://github.com/user-attachments/assets/8a166663-e0c8-4790-b588-b47d855a297d)

Test case here, all ways lead to toogle playlist. Modifiers somehow doesn't work.

adipose committed:
support nFlags for X/Middle/Wheel

adipose commented:
> Test case here, all ways lead to toogle playlist. Modifiers somehow doesn't work.

Should be fixed, nFlags wasn't being passed.

xLn2 commented:
There is default key assing for long jump, if it is still a desired feature can be added in future.

![resim](https://github.com/user-attachments/assets/5055f765-8ee9-4dc8-b678-0b01b73d1a4b)

clsid2 commented on src/mpc-hc/MouseTouch.cpp:
```diff
@@ -423,11 +423,12 @@ void CMouse::InternalOnLButtonUp(UINT nFlags, const CPoint& point)
 void CMouse::InternalOnMButtonDown(UINT nFlags, const CPoint& point)
 {
     SetCursor(nFlags, point);
+    return;
```
?

adipose replied:
The old code supported mouse down and mouse up, I believe.  MPC-BE is working off "up" action, so I disabled these calls.

clsid2 replied:
Maybe instead of adding a return, comment out the skipped line. Then it is more obvious that it is intentional change for others who read the code. Otherwise it might look like a coding error.

clsid2 commented on src/mpc-hc/MouseTouch.cpp:
```diff
@@ -459,11 +460,12 @@ void CMouse::InternalOnRButtonDblClk(UINT nFlags, const CPoint& point)
 bool CMouse::InternalOnXButtonDown(UINT nFlags, UINT nButton, const CPoint& point)
 {
     SetCursor(nFlags, point);
+    return false;
```
?

adipose commented:
> There is default key assing for long jump, if it is still a desired feature can be added in future.
> 
> ![resim](https://private-user-images.githubusercontent.com/3073922/355435526-5055f765-8ee9-4dc8-b678-0b01b73d1a4b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI5NjM2NDYsIm5iZiI6MTcyMjk2MzM0NiwicGF0aCI6Ii8zMDczOTIyLzM1NTQzNTUyNi01MDU1Zjc2NS04ZWU5LTRkYzgtYjY3OC0wYjAxYjczZDFhNGIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgwNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MDZUMTY1NTQ2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MjkyNGJlNzE3YmM3NjlhN2E2YzYxMTFjMWFlYTg2MDc1ZTM2ZjQ0NjEzZmViODY4YTA5NzNjZWUzMzZiM2VmZiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.V7bl4C7rprf9ZablwHjtVtedjw-At8KVxa-rKuRfcJw)

I need to review all the options in the dropdown to make sure they map to real shortcuts.

adipose committed:
avoid context menu when right click + mouse action occured

adipose committed:
add large jump to default commands

adipose committed:
themed resizable dialog for adding mouse commands

clsid2 commented:
Can you make a fresh test build?

adipose committed:
Merge remote-tracking branch 'upstream/develop' into patch419

adipose commented:
Test build: https://mega.nz/file/0JYwTCRD#TQB2uUrLTsZvkH0NIwwsPGTNdyHd3NCfZOR1MqdtwlA

xLn2 commented:
![resim](https://github.com/user-attachments/assets/c65b53bd-ae9e-4a55-bbb6-6ad6012581db) 
![resim](https://github.com/user-attachments/assets/56e684eb-a7e1-49fe-ac53-d040ce98681a)


- Using Default button remove assign of Large jumps as default value.
- Click and Double click comboboxes decolorized by using Default button.
- Dropdown scrollbar overlaps texts. It is also seen on Keys Page and Fullscreen Page dropdowns.

Functionally, there seems to be no problem with its operation.

adipose committed:
comment out instead of return on DOWN commands

adipose commented:
>Using Default button remove assign of Large jumps as default value.

I don't understand.

xLn2 commented:
`Default` reset standard value for _**Ctrl + WheelUp = Jump Forward (large)**_ and _**Ctrl + WheelDown = Jump Backward (large)**_

adipose commented:
Not sure what you mean by standard value.  Defaults come from MPC-BE.

![image](https://github.com/user-attachments/assets/1b4918df-2239-41e7-8837-5dcb834c0ec7)

Is there a default for `CTRL+` ?

adipose committed:
redraw mouse combos on change

adipose commented:
>Click and Double click comboboxes decolorized by using Default button.

Fixed.

adipose committed:
fix combobox widths

adipose commented:
>Dropdown scrollbar overlaps texts. It is also seen on Keys Page and Fullscreen Page dropdowns.

Fixed.

xLn2 commented:
> There is default key assing for long jump, if it is still a desired feature can be added in future.
> 
> ![resim](https://private-user-images.githubusercontent.com/3073922/355435526-5055f765-8ee9-4dc8-b678-0b01b73d1a4b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjMxNDc2NjUsIm5iZiI6MTcyMzE0NzM2NSwicGF0aCI6Ii8zMDczOTIyLzM1NTQzNTUyNi01MDU1Zjc2NS04ZWU5LTRkYzgtYjY3OC0wYjAxYjczZDFhNGIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgwOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MDhUMjAwMjQ1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OWVhNTIwMTFjZDdiYWMxZDg3YzBiMzNiZWZjYjI4ZTEyNmQ4OGNlM2NmZWUzMzQ1MTJhZTlmYzVhN2NlZTI1MyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.r5HlhiqVxPxymSpTH8yUakRQ55ISXs9eg8k_JNpLwBk)

It is currently default value for Mpc-hc.

adipose committed:
fix regression

adipose committed:
support mpchc defaults for ctrl+wheel

adipose commented:
> > There is default key assing for long jump, if it is still a desired feature can be added in future.
> > ![resim](https://private-user-images.githubusercontent.com/3073922/355435526-5055f765-8ee9-4dc8-b678-0b01b73d1a4b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjMxNDc2NjUsIm5iZiI6MTcyMzE0NzM2NSwicGF0aCI6Ii8zMDczOTIyLzM1NTQzNTUyNi01MDU1Zjc2NS04ZWU5LTRkYzgtYjY3OC0wYjAxYjczZDFhNGIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgwOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MDhUMjAwMjQ1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9OWVhNTIwMTFjZDdiYWMxZDg3YzBiMzNiZWZjYjI4ZTEyNmQ4OGNlM2NmZWUzMzQ1MTJhZTlmYzVhN2NlZTI1MyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.r5HlhiqVxPxymSpTH8yUakRQ55ISXs9eg8k_JNpLwBk)
> 
> It is currently default value for Mpc-hc.

I've made it the default, now.  Was it showing before you clicked default?  Because I don't think it would.

xLn2 commented:
> I've made it the default, now. Was it showing before you clicked default? Because I don't think it would.

When I first opened the app I got the wrong impression that it was. I probably set it up myself and forgot about it. Sorry for misinformation.

adipose commented:
Another test build:

https://mega.nz/file/ZZ5FmJ4b#cQE1CaUvABdDfi_CwvcOy9jo5lWMAbYjhGBWLb6bbhE

xLn2 commented:
3rd Test build:

![resim](https://github.com/user-attachments/assets/ddd4a639-78e2-49e3-a0c8-a04b1ba291ed)

Jump large shortcut's keyboard keys seems removed
Keys page still shows `Mouse Modifier` and `Mouse` column, maybe it is things to do after this patch merged, I have no idea.

I couldn't find anything else, thanks for your efforts.

adipose commented:
> Jump large shortcut's keyboard keys seems removed 

You must have been playing around with the "reset selected" because that key doesn't have a default and hasn't for some time (here's a 6 month old commit where you can see there was no default):

https://github.com/clsid2/mpc-hc/commit/d32a4483f059788f33a2b8457f89ec098e88b145#diff-e262625b6e0398aaa4b2cd3d48e671a0f9a90a0dca0a559682044d4c6a26fd05

![image](https://github.com/user-attachments/assets/a1eb0b98-a25b-46ab-a3f4-5237118e8f77)


>Keys page still shows `Mouse Modifier` and `Mouse` column, maybe it is things to do after this patch merged, I have no idea.

It can be removed, it's just a bit of an effort as it effects every single shortcut.  It's not active though.

adipose committed:
.

clsid2 commented:
Adjusting keys page is trivial and can be done after merge.

I will test new build soon.

clsid2 commented:
- There is an empty row in bottom of table that result in a vertical scrollbar being shown
- I suggest moving registry settings from `HKEY_CURRENT_USER\Software\MPC-HC\MPC-HC\Settings\Mouse` to `HKEY_CURRENT_USER\Software\MPC-HC\MPC-HC\Mouse`
- Registry values `ButtonLeftOpenRecent` and `EasyMove` are obsolete can be removed. Don't forget removing the added resource ids, etc.
- Set default for Right+Wheel Up/Down to jump forward/backward (large). Same for Ctrl+Wheel Up/Down.
- Default width of Shift and Control columns could be a bit bigger. They are narrow when containing only empty values.

After this it should be ready to merge.

clsid2 commented on src/mpc-hc/PPageSheet.h:
```diff
@@ -94,6 +95,7 @@ class CPPageSheet : public TreePropSheet::CTreePropSheet, public CMPCThemeUtil
     CPPageTheme m_theme;
     CPPageFormats m_formats;
     CPPageAccelTbl m_acceltbl;
+	CPPageMouse           m_mouse;
```
fix whitespace

adipose committed:
fix formatting

adipose commented:
>I suggest moving registry settings from HKEY_CURRENT_USER\Software\MPC-HC\MPC-HC\Settings\Mouse to HKEY_CURRENT_USER\Software\MPC-HC\MPC-HC\Mouse

If future settings are segregated, would we also top-level them and ditch the settings section?

adipose commented:
>There is an empty row in bottom of table that result in a vertical scrollbar being shown

For me there is no scrollbar.  It seems like it should be the opposite?  If there is an extra empty row then why have a scrollbar?

adipose committed:
.

clsid2 commented:
> If future settings are segregated, would we also top-level them and ditch the settings section?

Moving existing settings would break any apps they depend on their current location. For example Madvr reads certain settings.

clsid2 commented:
With your test build:
![mpc_mouse_test2](https://github.com/user-attachments/assets/4841ea7a-216e-4dc3-b7c4-23f4e50caa71)

adipose committed:
bigger listview

adipose commented:
I don't know why, but at 100% dpi it is too small and at larger DPI it's too big. But it should be big enough now.

adipose committed:
remove unused mouse settings

adipose commented:
>Set default for Right+Wheel Up/Down to jump forward/backward (large). Same for Ctrl+Wheel Up/Down.

Was already done for CTRL.  What about shift, no?

adipose committed:
.

adipose commented:
I think all those features are done now.

clsid2 commented:
Can you rebase? Github shows a merge conflict.

adipose committed:
merge upstream

clsid2 commented:
Issue I noticed:
Mouse pointer auto-hides while right-click context menu is open.

adipose commented:
> Issue I noticed: Mouse pointer auto-hides while right-click context menu is open.

Are you talking about, when I right click on the FS window, if I come back to the mainview, the pointer goes behind the window?

clsid2 commented:
It happens when right-clicking in normal windowed view. But yes, it seems to go behind the window. You also need to toggle to separate fullscreen and back first to trigger it.

adipose commented:
> It happens when right-clicking in normal windowed view. But yes, it seems to go behind the window. You also need to toggle to separate fullscreen and back first to trigger it.

I couldn't reproduce that but I can while FS is still active.

bibsp commented:
I have a question to adipose, before clsid release another test build.
Have you moved mouse double click speed setting for advanced page to this page? That is also a mouse related setting and better to keep everything in one page.

In the click section, add a new option like `Play/Pause with delay`
If that option is selected, then a new option will be available to configure the delay amount. Make those drop-down menu boxes smaller and there will be enough free space at right side.

adipose commented:
Hmm, what's the use case for the delay?

bibsp commented:
Toggle full screen using mouse double click will not play/pause the video just like YouTube player.
This is not a feature request. It's already there im MPC-HC.
Just move the setting called `Mouse leftup delay` from advanced page to this page.

clsid2 commented:
Maybe in future. I still consider that an advanced option right now. Plenty of people did not like it when briefly being enabled in an old test build.

bibsp commented:
Yes, I remember. I'm not telling you to enable delay by default. I'm just considering everything in one place. Hope you will add it in future. Thanks. Bye.
